### PR TITLE
H211: TTA mirror-aug on H190 EP13 — own val recovery

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Your H190 (mirror-aug p=0.25) reached val_abupt=6.077% (PASSES val gate by 5.9bp), but test failed: test_WSS=6.820% (+6.8bp), with WSS slope FLATTEST recorded (−0.085pp) and WSS_x INVERTED to +0.035pp. Your analysis (PR #1364) cleanly mapped this: p=0.25 is BELOW the slope-mechanism threshold — val gain is val-set-specific over-regularization.

**But** — the model DID still learn partial mirror invariance during training (mirror_flip_fraction ≈ 0.257 averaged across batches). That partial invariance may be EXPLOITABLE at inference via TTA.

**Hypothesis**: Test-Time Augmentation (TTA) — averaging the model's prediction on ORIGINAL geometry with its prediction on MIRRORED geometry — can recover test_WSS by exploiting the learned (partial) mirror invariance. If TTA narrows the val→test gap from −0.085pp to ≥−0.20pp, test_WSS may clear gate.

The 3-TTA-sprint cohort (alphonse H206 on H183, frieren H209 on H185, nezuko H211 on H190 — THIS) directly compares TTA recovery across the 3 val-SOTA runs that ALL fail test gates. Each was trained with mirror-aug but at different intensities (H183 p=0.5+tau_y=3.0, H185 p=0.5+GradNorm, H190 p=0.25).

**Comparison framing**:
- If H211 (p=0.25 + TTA) shows LARGER TTA gain than H206/H209 → confirms p=0.25 has UNDER-utilized mirror invariance that TTA recovers
- If H211 shows SMALLER TTA gain → confirms p=0.25 has LESS mirror invariance learned (below threshold)
- Mechanism attribution either way

## Instructions

**This is a pure eval sprint — NO training. No source file modifications to `data/loader.py` (read-only).**

### Step 1: Locate H190 EP13 checkpoint (your own run)
- W&B run: `9f2jtrg2` (project: `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`)
- W&B artifact: `model-nezuko-h190-mirror-aug-p025-9f2jtrg2` (alias `best`/`epoch-13`)
- Download to local disk

### Step 2: TTA mirror-aug pipeline

For each test mesh:
1. **Pass A**: forward pass on ORIGINAL geometry (no input transformation)
2. **Pass B**: forward pass on **left-right mirrored** input — apply the EXACT same mirror logic used in training (you implemented this in `data/loader.py`):
   - Negate y in `surface_x` (positions)
   - Negate normal_y in `surface_x` (normals)
   - Negate y in `volume_x`
   - INVARIANT (do NOT negate): cp, tau_x, tau_z, sdf, volume_pressure
3. **Un-mirror Pass B outputs**: WSS_y output needs sign flipped back. Verify the output spec.
4. **Average**: pred_final = 0.5 × pred_A + 0.5 × pred_B

If `train.py --eval_only` doesn't natively support TTA, implement a thin Python wrapper. You're well-positioned for this since you authored the mirror-aug code in `data/loader.py`.

### Step 3: Three eval runs

```bash
python -m torch.distributed.run --nproc-per-node=8 \
  train.py --eval_only \
  --checkpoint <path-to-h190-ep13> \
  --wandb_group h211-nezuko-tta-h190 \
  --run_name h211-h190-original

# repeat with mirrored-only and TTA-averaged modes
```

### Step 4: Report

For each eval mode (A=original, B=mirrored, C=TTA):
- test_WSS, test_WSS_x, test_WSS_y, test_WSS_z, test_VP, test_SP
- val_abupt sanity check (should reproduce 6.077%)
- Delta C vs A (TTA gain on test) — compare to your H190 vs H148 cohort framing
- WSS_x slope sign (was +0.035pp at terminal — does TTA restore to negative?)

### Step 5: Merge candidate

Criterion: test_WSS ≤ 6.752% AND test_VP ≤ 3.421% AND test_SP ≤ 3.577% AND val_abupt < 6.1358% AND WSS_x slope ≤ −0.020pp.

If TTA recovers WSS_x to negative AND test_WSS clears gate → SOTA candidate, tag advisor.

## Baseline

Current best (H112, PR #1283, W&B run `u9ue2ryb`):
- val_abupt: **6.1358%**
- test_WSS: **6.752%**
- test_VP: **3.421%**
- test_SP: **3.695%** (strict gate ≤ 3.577%)

H190 terminal (your reference):
- val_abupt = 6.077% (val gate PASS)
- test_WSS = 6.820% (+6.8bp above gate)
- WSS_x slope = +0.035pp (basin INVERTED)
- WSS slope = −0.085pp (FLATTEST in cohort)

## Coordination

- alphonse H206 = TTA on H183 EP13 (val 6.039%)
- frieren H209 = TTA on H185 EP13 (val 6.017%)
- nezuko H211 = THIS — TTA on H190 EP13 (val 6.077%)

These 3 experiments cleanly cross-validate the mirror-invariance-recovery hypothesis across 3 different mirror-aug recipes.

Expected runtime: ~1h on 8 GPUs (3 evals × ~20min each).

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT nezuko:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["vdub0ga5"],"primary_metric":{"name":"full_val/tta/val_surface/abupt_axis_mean_rel_l2_pct","value":6.0363},"test_metric":{"name":"test/tta/test_surface/wall_shear_rel_l2_pct","value":6.7800}}

## Results

TTA mirror-aug on H190 EP13 (W&B run `vdub0ga5`, source ckpt `model-nezuko-h190-mirror-aug-p025-9f2jtrg2:best`, source epoch=13, source=ema). DDP-sharded full eval on 8 GPUs (val 132.7s / test 198.8s); peak ~16 GB/GPU.

**Val sanity (original mode):** `abupt_axis_mean_rel_l2_pct` = **6.0764%** ✓ reproduces H190 terminal val 6.077% to 4dp.

### Three-mode table (all `_rel_l2_pct`)

#### Val
| Mode      | abupt   | WSS    | WSS_x  | WSS_y  | WSS_z  | SP     | VP     |
|-----------|---------|--------|--------|--------|--------|--------|--------|
| original  | 6.0764  | 6.9047 | 6.0341 | 7.5514 | 9.2925 | 4.0106 | 3.4936 |
| mirrored  | 6.0928  | 6.9258 | 6.0505 | 7.5920 | 9.3068 | 4.0198 | 3.4949 |
| **tta**   | **6.0363** | **6.8617** | **6.0073** | **7.4659** | **9.2474** | **3.9860** | **3.4747** |

#### Test
| Mode      | abupt   | WSS    | WSS_x  | WSS_y  | WSS_z  | SP     | VP     |
|-----------|---------|--------|--------|--------|--------|--------|--------|
| original  | 5.8994  | 6.8198 | 6.0691 | 7.3899 | 8.8266 | 3.7399 | 3.4713 |
| mirrored  | 5.9196  | 6.8416 | 6.0870 | 7.4364 | 8.8331 | 3.7553 | 3.4862 |
| **tta**   | **5.8639** | **6.7800** | **6.0445** | **7.3139** | **8.7815** | **3.7196** | **3.4599** |

### TTA gain on test (TTA − original, pp)

| metric   | delta |
|----------|-------|
| abupt    | −0.0355 |
| WSS      | −0.0398 |
| WSS_x    | −0.0246 |
| WSS_y    | **−0.0760** |
| WSS_z    | −0.0451 |
| SP       | −0.0203 |
| VP       | −0.0114 |

Mirrored-only (pass B alone) is a uniform regression of ≈ +0.02pp across the board → confirms the un-mirror logic is correct (B is on the same scale as A, just slightly worse) and the TTA average is genuinely doing better than either component.

### Val → test slope check (test − val, pp)

| metric | original | mirrored | tta |
|--------|---------:|---------:|----:|
| WSS    | −0.0849 (PR-body match: −0.085) | −0.0842 | −0.0817 |
| WSS_x  | **+0.0350** (PR-body match: +0.035) | +0.0365 | **+0.0372** |
| abupt  | −0.1770 | −0.1732 | −0.1724 |

WSS_x slope stays INVERTED under TTA (+0.037pp, even slightly worse than +0.035pp original). TTA narrows the WSS basin slope from −0.085 to −0.082 (≈ +0.003pp flattening), so the val→test gap shrinks only marginally. TTA acts ≈ uniformly across val & test — it does not selectively recover test.

### Merge-gate check (test mode = tta vs current best H112 u9ue2ryb)

| gate | TTA | required | delta vs gate |
|------|-----|----------|---------------|
| test_WSS  | 6.7800 | ≤ 6.752 | **+0.028 FAIL** |
| test_VP   | 3.4599 | ≤ 3.421 | **+0.039 FAIL** |
| test_SP   | 3.7196 | ≤ 3.577 (strict) | **+0.143 FAIL** |
| val_abupt | 6.0363 | < 6.1358 | −0.099 PASS |
| WSS_x slope | +0.037pp | ≤ −0.020pp | **+0.057 FAIL** |

**Not a SOTA / merge candidate.** TTA narrows the val→test WSS gap by only +0.003pp (−0.085 → −0.082), nowhere near the +0.065pp swing needed to clear gate, and WSS_x slope stays positive.

### Command
```bash
torchrun --standalone --nproc-per-node=8 tta_eval.py \
  --checkpoint outputs/h211_tta/h190_ep13_best/checkpoint.pt \
  --wandb-group h211-nezuko-tta-h190 \
  --wandb-name nezuko/h211-tta-h190-ep13
```

W&B run: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/vdub0ga5

### What happened

H190 trained with p=0.25 mirror-aug DID learn partial mirror invariance, and TTA does recover some of it: averaging original-input and mirrored-input predictions cuts test_WSS by −0.040pp and test_WSS_y (the channel that flips under reflection) by −0.076pp — the largest single-channel gain, exactly where the mechanism predicts. WSS_z gets −0.045pp (sanity — z is invariant under y-flip but the model has noise, so averaging still helps). The order WSS_y > WSS_z > WSS_x in TTA gain magnitude is consistent with "the model has more residual disagreement between original and mirrored on the channel whose sign actually flips."

But the gain is small in absolute terms (−0.040pp on WSS), nothing like the +0.065pp needed to clear the test gate. The val→test gap stays at −0.082pp under TTA (vs −0.085pp original) — TTA acts roughly uniformly on val & test rather than selectively closing the gap, which is consistent with H190's terminal characterization: p=0.25 is below the slope-mechanism threshold, the gap is structural (val-set-specific over-regularization at low p), not a recoverable TTA-shaped deficit.

Cohort comparison (advisor's framing): H211 TTA gain on test_WSS = −0.040pp. Mechanism reading: p=0.25 has UNDER-utilized mirror invariance (the model didn't fully learn it during training because p was too low), but TTA can only recover what the model partially knows — so the recovery is small. Compare to H206 (p=0.5 + tau_y=3.0) and H209 (p=0.5 + GradNorm) which should show LARGER TTA gains if they actually internalised mirror invariance more strongly.

Mirrored-only test (pass B alone): regresses ≈ +0.02pp uniformly across all metrics → confirms (a) the un-mirror sign-flip on tau_y is correct, (b) the mirrored input is in a slightly-out-of-distribution regime for this p=0.25 model (because it never saw a guaranteed-mirrored input at train time — only 25% of batches got the flip), so the model is slightly less confident on the mirrored frame. TTA recovers the loss by averaging out the per-frame disagreement.

### Suggested follow-ups

1. **Cohort cross-eval (already in-flight):** the comparison against alphonse H206 (TTA on H183, p=0.5+tau_y=3.0) and frieren H209 (TTA on H185, p=0.5+GradNorm) is the load-bearing comparison; H211 alone doesn't conclude — the three together tell us whether higher p actually buys TTA-recoverable invariance.
2. **Weighted TTA:** 0.5 / 0.5 averaging is the simplest form. If pass A is generally better than pass B (mirrored-only is +0.02pp regression), a learned or fixed weight α with pred = (1−α) A + α B and α < 0.5 could squeeze a bit more out (val sweep would say, ~0.02–0.05 of additional gain at most).
3. **TTA on a model that PASSED test gates (e.g. H112 baseline):** sanity check — does TTA on a model trained without mirror-aug also show a tiny gain (consistent with averaging noise), or specifically nothing? Distinguishes "TTA improvement is mirror-invariance-specific" from "TTA averaging always helps a bit."
4. **TTA on higher-p models (>0.5):** would require new training; orthogonal to this sprint. The H206/H209 results bound this.

This was a clean negative result for the merge candidacy and a clean positive for the mechanism attribution (mirror_y_TTA_gain is the largest channel, as predicted by reflection symmetry of tau_y).

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1384 because ADVISOR: **H211 closed — clean negative on test_WSS gate. Banking as part of Finding Q.**

Excellent execution. Eval pipeline integrity confirmed.

### H211 result snapshot

| Metric | orig (H190 EP13) | TTA | gate |
|---|---:|---:|---|
| val_abupt | 6.077 | 6.0363 | < 6.1358 ✓ PASS |
| test_WSS | ~6.82 | **6.7800** | ≤ 6.752 ✗ FAIL by +2.8bp |
| WSS_x slope | flipped | +0.037 | NOT recovered |

H190 had higher pre-TTA test_WSS (~6.82) than H185 (6.7642), so the −4bp TTA gain wasn't enough to clear the gate.

### Finding Q (program-permanent): TTA on mirror-aug-trained models

See PR #1379 (alphonse H206 close) for the full table across all 3 TTA recovery sprints.

TL;DR:
1. **TTA gives ~4-5bp uniform gain on test_WSS.** Worth applying to any future mirror-aug winner at inference.
2. **WSS_x slope is NOT recovered by TTA.** Basin disruption is a weight-space training-time property, not a test-time prediction-symmetry artifact.
3. **H185+TTA (frieren H209, PR #1382) is the merged winner** — it had the lowest test_WSS pre-TTA (6.7642), so the TTA gain crossed the gate at 6.7214.

### Cross-validation with H210 (tanjiro)

Your H211 result corroborates Finding O (cross-recipe SWA destroys models) and Finding P (basins not linearly mode-connected) by showing that the **WSS_x basin flip is a training-trajectory property**, not a representational/symmetry property at evaluation time. The three findings together close two large arms (weight-averaging, TTA recovery) cleanly.

### Next assignment

Branch incoming for H216 — pivoting from TTA recovery to a different direction.

H185+TTA is now merged SOTA. The remaining recovery arms are: (a) block-wise structural splicing (tanjiro H213 active), (b) sub-alpha micro-blends (askeladd H214 active), (c) fine-tune-from-H112-with-anchor (untried, your candidate), (d) per-channel head distillation. Your next sprint will be one of (c)/(d).

Closing this PR. Result logged. Banking Finding Q..
